src/cupbearer/detectors/finetuning.py

```diff
@@ -0,0 +1,145 @@
+from dataclasses import dataclass, field
+
+import jax
+import jax.numpy as jnp
+import optax
+from torch.utils.data import DataLoader
+
+from cupbearer.data import numpy_collate
+from cupbearer.detectors.anomaly_detector import AnomalyDetector
+from cupbearer.detectors.config import DetectorConfig, TrainConfig
+from cupbearer.utils import trainer
+from cupbearer.utils.config_groups import config_group
+from cupbearer.utils.optimizers import Adam, OptimizerConfig
+
+
+class FinetuningTrainer(trainer.TrainerModule):
+    def __init__(self, model, loss_fn, **kwargs):
+        super().__init__(model=model, **kwargs)
+        self.loss_fn = loss_fn
+
+    def create_functions(self):
+        def train_step(state, batch):
+            def loss_fn(params):
+                return self.loss_fn(params, batch)
+
+            loss, grads = jax.value_and_grad(loss_fn)(state.params)
+            state = state.apply_gradients(grads=grads)
+            metrics = {"loss": loss}
+            return state, metrics
+
+        def eval_step(state, batch):
+            loss = self.loss_fn(state.params, batch)
+            metrics = {"loss": loss}
+            return metrics
+
+        return train_step, eval_step
+
+
+class FinetuningAnomalyDetector(AnomalyDetector):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.finetuned_params = None
+
+    def train(
+        self,
+        clean_dataset,
+        optimizer: OptimizerConfig,
+        num_epochs: int = 10,
+        batch_size: int = 128,
+        # Not used, but will be passed by the detector train script
+        debug: bool = False,
+    ):
+        trainer_instance = FinetuningTrainer(
+            model=self.model,
+            loss_fn=self.loss_fn,
+            optimizer=optimizer.build(),
+            log_dir=self.save_path,
+            override_variables={
+                "params": self.params
+            },  # Use existing parameters for initialization
+            example_input=clean_dataset[0][0][None],
```

Comment:
In case of image classification `clean_dataset[0][0][None]` is first image and add an extra last dimension (channel dimension?). It runs so it's probably correct, but I would have expected `clean_dataset[0][0][None, ...]` to be what we wanted?

Reply:
`[None]` is equivalent to `[None, ...]`, it adds a singleton dimension at the front, not the back. That's the batch dimension (which we removed by indexing into the first batch before). But I agree your version is more readable and I should also add a comment explaining all the indices here.